File: src/main/java/destiny/iching/HexagramBoShuComparator.java
/**
 * @author smallufo
 * Created on 2011/1/24 at 下午4:00:08
 */
package destiny.iching;

import com.google.common.collect.BiMap;
import com.google.common.collect.ImmutableBiMap;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;

/**
 * 帛書卦序
 */
public class HexagramBoShuComparator extends AbstractHexagramSequence implements Comparator<HexagramIF> {

  private static final BiMap<Hexagram,Integer> map = new ImmutableBiMap.Builder<Hexagram , Integer>()
    .put(Hexagram.乾, 1)
    .put(Hexagram.否, 2)
    .put(Hexagram.遯, 3)
    .put(Hexagram.履, 4)
    .put(Hexagram.訟, 5)
    .put(Hexagram.同人, 6)
    .put(Hexagram.無妄, 7)
    
    .put(Hexagram.姤, 8)
    .put(Hexagram.艮, 9)
    .put(Hexagram.大畜, 10)
    .put(Hexagram.剝, 11)
    .put(Hexagram.損, 12)
    
    .put(Hexagram.蒙, 13)
    .put(Hexagram.賁, 14)
    .put(Hexagram.頤, 15)
    .put(Hexagram.蠱, 16)
    .put(Hexagram.坎, 17)
    
    .put(Hexagram.需, 18)
    .put(Hexagram.比, 19)
    .put(Hexagram.蹇, 20)
    .put(Hexagram.節, 21)
    .put(Hexagram.既濟, 22)
    
    .put(Hexagram.屯, 23)
    .put(Hexagram.井, 24)
    .put(Hexagram.震, 25)
    .put(Hexagram.大壯, 26)
    .put(Hexagram.豫, 27)
    
    .put(Hexagram.小過, 28)
    .put(Hexagram.歸妹, 29)
    .put(Hexagram.解, 30)
    
    .put(Hexagram.豐, 31)
    .put(Hexagram.恆, 32)
    .put(Hexagram.坤, 33)
    .put(Hexagram.泰, 34)
    
    .put(Hexagram.謙, 35)
    .put(Hexagram.臨, 36)
    .put(Hexagram.師, 37)
    .put(Hexagram.明夷, 38)
    
    .put(Hexagram.復, 39)
    .put(Hexagram.升, 40)
    .put(Hexagram.兌, 41)
    .put(Hexagram.夬, 42)
    .put(Hexagram.萃, 43)
    .put(Hexagram.咸, 44)
    .put(Hexagram.困, 45)
    
    .put(Hexagram.革, 46)
    .put(Hexagram.隨, 47)
    .put(Hexagram.大過, 48)
    .put(Hexagram.離, 49)
    .put(Hexagram.大有, 50)
    .put(Hexagram.晉, 51)
    
    .put(Hexagram.旅, 52)
    .put(Hexagram.睽, 53)
    .put(Hexagram.未濟, 54)
    .put(Hexagram.噬嗑, 55)
    .put(Hexagram.鼎, 56)
    .put(Hexagram.巽, 57)
    
    .put(Hexagram.小畜, 58)
    .put(Hexagram.觀, 59)
    .put(Hexagram.漸, 60)
    .put(Hexagram.中孚, 61)
    
    .put(Hexagram.渙, 62)
    .put(Hexagram.家人, 63)
    .put(Hexagram.益, 64).build();

  @Override
  protected BiMap<Hexagram, Integer> getMap() {
    return map;
  }

  @Override
  public int compare(@NotNull HexagramIF h1, @NotNull HexagramIF h2)
  {
    return getIndex(h1)-getIndex(h2);
  }

}

File: src/main/java/destiny/core/iching/HexagramBoShuComparator.kt
/**
 * @author smallufo
 * Created on 2011/1/24 at 下午4:00:08
 */
package destiny.core.iching

/**
 * 帛書卦序
 */
class HexagramBoShuComparator : AbstractHexagramSequence(), Comparator<IHexagram> {

  override val map: Map<Hexagram, Int>
    get() = mapOf(
      Hexagram.乾 to 1 ,
      Hexagram.否 to 2 ,
      Hexagram.遯 to 3 ,
      Hexagram.履 to 4 ,
      Hexagram.訟 to 5 ,
      Hexagram.同人 to 6 ,
      Hexagram.無妄 to 7 ,
      Hexagram.姤 to 8 ,
      Hexagram.艮 to 9 ,
      Hexagram.大畜 to 10 ,
      Hexagram.剝 to 11 ,
      Hexagram.損 to 12 ,
      Hexagram.蒙 to 13 ,
      Hexagram.賁 to 14 ,
      Hexagram.頤 to 15 ,
      Hexagram.蠱 to 16 ,
      Hexagram.坎 to 17 ,
      Hexagram.需 to 18 ,
      Hexagram.比 to 19 ,
      Hexagram.蹇 to 20 ,
      Hexagram.節 to 21 ,
      Hexagram.既濟 to 22 ,
      Hexagram.屯 to 23 ,
      Hexagram.井 to 24 ,
      Hexagram.震 to 25 ,
      Hexagram.大壯 to 26 ,
      Hexagram.豫 to 27 ,
      Hexagram.小過 to 28 ,
      Hexagram.歸妹 to 29 ,
      Hexagram.解 to 30 ,
      Hexagram.豐 to 31 ,
      Hexagram.恆 to 32 ,
      Hexagram.坤 to 33 ,
      Hexagram.泰 to 34 ,
      Hexagram.謙 to 35 ,
      Hexagram.臨 to 36 ,
      Hexagram.師 to 37 ,
      Hexagram.明夷 to 38 ,
      Hexagram.復 to 39 ,
      Hexagram.升 to 40 ,
      Hexagram.兌 to 41 ,
      Hexagram.夬 to 42 ,
      Hexagram.萃 to 43 ,
      Hexagram.咸 to 44 ,
      Hexagram.困 to 45 ,
      Hexagram.革 to 46 ,
      Hexagram.隨 to 47 ,
      Hexagram.大過 to 48 ,
      Hexagram.離 to 49 ,
      Hexagram.大有 to 50 ,
      Hexagram.晉 to 51 ,
      Hexagram.旅 to 52 ,
      Hexagram.睽 to 53 ,
      Hexagram.未濟 to 54 ,
      Hexagram.噬嗑 to 55 ,
      Hexagram.鼎 to 56 ,
      Hexagram.巽 to 57 ,
      Hexagram.小畜 to 58 ,
      Hexagram.觀 to 59 ,
      Hexagram.漸 to 60 ,
      Hexagram.中孚 to 61 ,
      Hexagram.渙 to 62 ,
      Hexagram.家人 to 63 ,
      Hexagram.益 to 64 )

  override fun compare(h1: IHexagram, h2: IHexagram): Int {
    return getIndex(h1) - getIndex(h2)
  }


}
